ownCloud/Client/Actions/Scanner/ScanViewController.swift

```diff
@@ -0,0 +1,526 @@
+			pageCell.page = pages[indexPath.item]
+		}
+
+		return cell
+	}
+}
+
+extension CGRect {
+	func asData() -> NSData {
+		return NSData(from: self)
+	}
+}
+
+struct ScanExportFormat {
+	var name : String
+	var suffix : String
+	var supportsMultiPage : Bool = false
+	var exporter : ((_ baseURL: URL, _ pages: [ScanPage]) -> ([URL]?))?
+}
+
+@available(iOS 13.0, *)
+class ScanViewController: StaticTableViewController {
+	weak var core : OCCore?
+	var targetFolderItem : OCItem?
+
+	// Pages
+	var pagesSection : StaticTableViewSection
+	var pagesCollectionViewController : ScanPagesCollectionViewController?
+
+	// Save
+	var saveSection : StaticTableViewSection
+	var fileNameRow : StaticTableViewRow?
+
+	// Options
+	var optionsSection : StaticTableViewSection
+	var formatSegmentedControl : UISegmentedControl?
+	var oneFilePerPageRow : StaticTableViewRow?
+
+	private static func exportAsImage(format: String, baseURL: URL, pages: [ScanPage]) -> [URL]? {
+		var files : [URL] = []
+
+		for page in pages {
+			let targetURL = baseURL.appendingPathComponent(UUID().uuidString)
+			var imageData : Data?
+
+			switch format {
+				case "jpeg":
+					imageData = page.image.jpegData(compressionQuality: 0.75)
+
+				case "png":
+					imageData = page.image.pngData()
+
+				default: break
+			}
+
+			if let imageData = imageData {
+				if (try? imageData.write(to: targetURL)) != nil {
+					// Writing successful
+					files.append(targetURL)
+				} else {
+					// Error writing
+					return nil
+				}
+			} else {
+				// Error
+				return nil
+			}
+		}
+
+		return files.count > 0 ? files : nil
+	}
+
+	private static func exportAsPDF(pages: [ScanPage], to url: URL) -> Bool {
+		if let pdfContext = CGContext(url as CFURL, mediaBox: nil, nil) {
+			defer {
+				pdfContext.closePDF()
+			}
+
+			for page in pages {
+				if let jpegData = page.image.jpegData(compressionQuality: 0.75),
+				   let dataProvider = CGDataProvider(data: jpegData as CFData),
+				   let jpegImage = CGImage(jpegDataProviderSource: dataProvider, decode: nil, shouldInterpolate: true, intent: .defaultIntent) {
+					let mediaBoxRect = CGRect(x: 0, y: 0, width: page.image.size.width, height: page.image.size.height)
+
+					pdfContext.beginPDFPage([
+						kCGPDFContextMediaBox : mediaBoxRect.asData()
+					] as CFDictionary)
+
+					pdfContext.draw(jpegImage, in: mediaBoxRect)
+
+					pdfContext.endPDFPage()
+				} else {
+					return false
+				}
+			}
+
+			return true
+		}
+
+		return false
+	}
+
+	let availableExportFormats : [ScanExportFormat] = [
+		ScanExportFormat(name: "PDF", suffix: "pdf", supportsMultiPage: true, exporter: { (baseURL, pages) in
+			let targetURL = baseURL.appendingPathComponent(UUID().uuidString)
+
+			if ScanViewController.exportAsPDF(pages: pages, to: targetURL) {
+				// Writing successful
+				return [targetURL]
+			} else {
+				// Error writing
+				return nil
+			}
+
+			// PDFKit implementation (~ 11x larger files (31 MB vs. 2.7 MB for a two page scan)):
```

Comment:
@felix-schwarz May be worth removing dead code since and just leave a remark why you dropped PDFKit based implementation?

Reply:
Good idea. I removed that code and placed an extended version of that comment above the code using Quartz to generate the PDF.